R/rectangle.R

```diff
@@ -317,152 +729,123 @@ guess_dir <- function(x, col) {
+  ellipsis::check_dots_empty()
 
-    if (is.numeric(idx)) {
-      x[-idx]
-    } else if (is.character(idx)) {
-      if (has_name(x, idx)) {
-        x[setdiff(names(x), idx)]
-      } else {
-        x
-      }
-    } else {
-      x
-    }
-  } else {
-    cur_idx <- idx[[1]]
-    idx <- idx[-1]
+  if (!vec_is_list(ptype)) {
+    abort("`ptype` must be a list.")
+  }
+  if (!vec_is_list(ptype)) {
+    abort("`transform` must be a list.")
+  }
+  if (!is_bool(simplify)) {
+    abort("`simplify` must be a single `TRUE` or `FALSE`.")
+  }
 
-    if (is.numeric(cur_idx)) {
-      if (cur_idx > length(x)) {
-        x
-      } else {
-        x[[cur_idx]] <- strike(x[[cur_idx]], idx)
-        x
-      }
-    } else if (is.character(cur_idx)) {
-      if (!has_name(x, cur_idx)) {
-        x
-      } else {
-        x[[cur_idx]] <- strike(x[[cur_idx]], idx)
-        x
-      }
-    } else {
-      x
-    }
+  x_n <- length(x)
+  x_size <- vec_size(x)
+  x_names <- names(x)
+
+  out <- vector("list", length = x_n)
+  names(out) <- x_names
+
+  for (i in seq_len(x_n)) {
+    col <- x[[i]]
+    col_name <- x_names[[i]]
+
+    col_ptype <- ptype[[col_name]]
+    col_transform <- transform[[col_name]]
+
+    out[[i]] <- col_simplify(
+      x = col,
+      ptype = col_ptype,
+      transform = col_transform,
+      simplify = simplify
+    )
   }
+
+  new_data_frame(out, n = x_size)
 }
 
-simplify_col <- function(x, nm, ptype = list(), transform = list(), simplify = FALSE) {
-  transform <- transform[[nm]]
-  ptype <- ptype[[nm]]
+col_simplify <- function(x,
+                         ...,
+                         ptype = NULL,
+                         transform = NULL,
+                         simplify = TRUE) {
+  ellipsis::check_dots_empty()
 
   if (!is.null(transform)) {
-    x <- map(x, as_function(transform))
+    transform <- as_function(transform)
   }
 
-  if (!simplify || !is.list(x)) {
+  if (!vec_is_list(x)) {
+    if (!is.null(transform)) {
+      x <- transform(x)
+    }
+    if (!is.null(ptype)) {
+      x <- vec_cast(x, ptype)
+    }
     return(x)
   }
 
-  # Don't simplify lists of lists, because that typically indicates that
-  # there might be multiple values.
-  is_list <- map_lgl(x, is.list)
-  if (any(is_list)) {
-    if (is.null(ptype)) {
-      return(x)
-    } else {
-      abort(glue("Can't simplfy '{nm}'; contains a nested list"))
-    }
+  if (!is.null(transform)) {
+    x <- tidyr_new_list(x)
+    x <- map(x, transform)
+    # Can't convert result to list_of, as we can't be certain of element types
   }
-
-  # Don't try and simplify non-vectors
-  is_vec <- map_lgl(x, ~ vec_is(.x) || is.null(.x))
-  if (any(!is_vec)) {
-    if (is.null(ptype)) {
-      return(x)
-    } else {
-      abort(glue("Can't simplfy '{nm}'; contains a non-vector"))
-    }
+  if (!is.null(ptype)) {
+    x <- tidyr_new_list(x)
+    x <- vec_cast_common(!!!x, .to = ptype)
+    x <- new_list_of(x, ptype = ptype)
```

Comment:
I think this is pretty slick. It means that if you can't simplify (or choose not to), then you can create a list-of of the non-simplified results by supplying a ptype for that component. This was requested in https://github.com/tidyverse/tidyr/issues/998

``` r
library(tidyr)

df <- tibble(
  x = list(
    list(a = 1:2),
    list(a = 1)
  )
)

df
#> # A tibble: 2 × 1
#>   x               
#>   <list>          
#> 1 <named list [1]>
#> 2 <named list [1]>

unnest_wider(df, x)
#> # A tibble: 2 × 1
#>   a        
#>   <list>   
#> 1 <int [2]>
#> 2 <dbl [1]>

unnest_wider(df, x, ptype = list(a = integer()))
#> # A tibble: 2 × 1
#>             a
#>   <list<int>>
#> 1         [2]
#> 2         [1]
```

<sup>Created on 2021-11-08 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
That is cool!